### PR TITLE
Enable Windows Arm64

### PR DESCRIPTION
**Pull request by iremyux**
This PR enables Pytorch for Windows on Arm64 - CPU only. 
Currently, there aren't any checks in place to build and test for Windows on Arm64, but we're working to implement those as soon as possible. 
We recommend using [Arm Performance Libraries (APL)](https://developer.arm.com/Tools%20and%20Software/Arm%20Performance%20Libraries) as a BLAS option, which is introduced in this PR. 

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @Blackhex @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1142,6 +1142,7 @@ def main():
         "networkx",
         "jinja2",
         "fsspec",
+        'mkl>=2021.1.1,<=2021.4.0; platform_system == "Windows" and platform_machine != "ARM64"',
```
Comment (malfet) [inline setup.py:1145]: This feels like a wrong place to add the dependency, as Linux wheels might also depend on MKL if compiled with MKL, aren't they?

File: cmake/Modules/FindAPL.cmake [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+$ENV{ARMPL_DIR}/include
+)
+
+SET(APL_LIB_SEARCH_PATHS
+$ENV{ARMPL_DIR}/lib
+ )
+
+FIND_PATH(APL_INCLUDE_DIR NAMES armpl.h PATHS ${APL_INCLUDE_SEARCH_PATHS})
+FIND_LIBRARY(APL_LIB NAMES armpl_lp64.dll.lib PATHS ${APL_LIB_SEARCH_PATHS})
+FIND_LIBRARY(APL_MATH NAMES amath.dll.lib PATHS ${APL_LIB_SEARCH_PATHS})
```
Comment (malfet) [inline cmake/Modules/FindAPL.cmake:11]: Aren't APL also available for Linux? Is there a standard Cmake module available one can copy-n-pasted?

File: setup.py (hunk 1 above)
Comment (iremyux) [reply setup.py]: This was a mistake happened while rebasing. Since in pytorch/main mkl was removed, this change is not there anymore.

File: cmake/Modules/FindAPL.cmake (hunk 2 above)
Comment (nSircombe) [reply cmake/Modules/FindAPL.cmake:11]: Upstream CMake's version of FindBLAS - https://cmake.org/cmake/help/latest/module/FindBLAS.html

Comment (alinpahontu2912) [reply cmake/Modules/FindAPL.cmake:11]: Hey @malfet, I want to better understand the requirements. Is it okay to just update the FindAPL module to support linux too? @nSircombe thanks for the links, but it looks like the FindBlas/find_package imlementations do not recognize ArmPL on Windows currently: https://community.arm.com/support-forums/f/high-performance-computing-forum/56375/find_package-blas-fails-to-find-arm-performance-libraries/184602

File: aten/src/ATen/cpu/vec/vec_base.h [hunk 3]
```diff
@@ -290,13 +290,25 @@ struct Vectorized {
...
+  Vectorized<T> map(T (*const f)(T)) const {
+    Vectorized<T> ret;
+    for (int64_t i = 0; i < size(); i++) {
+      ret[i] = f(values[i]);
+      if (++i < size())
+        ret[i] = f(values[i]);
+    }
+    return ret;
+  }
+#else
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec_base.h:303]: As discussed, this feels unrelated, so please do it in separate PR
```suggestion
```

File: aten/src/ATen/cpu/vec/vec_base.h [hunk 4]
```diff
@@ -290,13 +290,25 @@ struct Vectorized {
...
   Vectorized<T> map(T (*const f)(T)) const {
     Vectorized<T> ret;
     for (int64_t i = 0; i != size(); i++) {
       ret[i] = f(values[i]);
     }
     return ret;
   }
+#endif
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec_base.h:311]:
```suggestion
```

File: aten/src/ATen/Config.h.in [hunk 5]
```diff
@@ -19,3 +19,4 @@
 #define AT_PARALLEL_NATIVE @AT_PARALLEL_NATIVE@
 #define AT_BLAS_F2C() @AT_BLAS_F2C@
 #define AT_BLAS_USE_CBLAS_DOT() @AT_BLAS_USE_CBLAS_DOT@
+#define AT_BUILD_WITH_APL() @AT_BUILD_WITH_APL@
```
Comment (malfet) [inline aten/src/ATen/Config.h.in:22]: Let's avoid this one
```suggestion
```

File: BUILD.bazel [hunk 6]
```diff
@@ -273,6 +273,7 @@ header_template_rule(
         "@AT_PARALLEL_NATIVE@": "1",
         "@AT_BLAS_F2C@": "0",
         "@AT_BLAS_USE_CBLAS_DOT@": "1",
+        "@AT_BUILD_WITH_APL@": "0",
```
Comment (malfet) [inline BUILD.bazel:276]:
```suggestion
```

File: aten/src/ATen/CMakeLists.txt [hunk 7]
```diff
@@ -34,6 +34,7 @@ set_bool(AT_MAGMA_ENABLED USE_MAGMA)
 set_bool(CAFFE2_STATIC_LINK_CUDA_INT CAFFE2_STATIC_LINK_CUDA)
 set_bool(AT_CUDNN_ENABLED CAFFE2_USE_CUDNN)
 set_bool(AT_CUSPARSELT_ENABLED CAFFE2_USE_CUSPARSELT)
+set_bool(AT_BUILD_WITH_APL APL_FOUND)
```
Comment (malfet) [inline aten/src/ATen/CMakeLists.txt:37]:
```suggestion
```

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 8]
```diff
@@ -132,11 +132,43 @@ extern "C" void dgetrf_(int *m, int *n, double *a, int *lda, int *ipiv, int *inf
 extern "C" void sgetrf_(int *m, int *n, float *a, int *lda, int *ipiv, int *info);
 
 // potrs
+#if defined(_WIN32) && defined(_M_ARM64)
+#if AT_BUILD_WITH_APL()
```
Comment (malfet) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:136]:
```suggestion
// TODO: Add reference to ARM Performance Library issue explaining why those are needed
```

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 9]
```diff
@@ -132,11 +132,43 @@ extern "C" void dgetrf_(int *m, int *n, double *a, int *lda, int *ipiv, int *inf
...
+
+static inline void dpotrs_(char *uplo, int *n, int *nrhs, double *a, int *lda, double *b, int *ldb, int *info){
+  *info = LAPACKE_dpotrs(LAPACK_COL_MAJOR, *uplo, *n, *nrhs, a, *lda, b, *ldb);
+}
+
+static inline void spotrs_(char *uplo, int *n, int *nrhs, float *a, int *lda, float *b, int *ldb, int *info) {
+  *info = LAPACKE_spotrs(LAPACK_COL_MAJOR, *uplo, *n, *nrhs, a, *lda, b, *ldb);
+}
+
+#endif
```
Comment (malfet) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:162]:
```suggestion
```

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 10]
```diff
@@ -284,11 +316,33 @@ extern "C" void dorgqr_(int *m, int *n, int *k, double *a, int *lda, double *tau
 extern "C" void sorgqr_(int *m, int *n, int *k, float *a, int *lda, float *tau, float *work, int *lwork, int *info);
 
 // ormqr
+#if AT_BUILD_WITH_APL()
```
Comment (malfet) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:319]:
```suggestion
#if defined(_WIN32) && defined(_M_ARM64)
// TODO: Explain why those are necessary
```

File: buckbuild.bzl [hunk 11]
```diff
@@ -1116,6 +1116,9 @@ def define_buck_targets(
             "--replace",
             "@AT_BLAS_USE_CBLAS_DOT@",
             "AT_BLAS_USE_CBLAS_DOT_FBXPLAT",
+            "--replace",
+            "@AT_BUILD_WITH_APL@",
+            "APL_FOUND",
```
Comment (malfet) [inline buckbuild.bzl:1121]:
```suggestion
```

**Comment by malfet** [comment]
@iremyux can you please allow contributions from reviewers so that I can apply review suggestions, rebase  and land the PR?

**Comment by iremyux** [comment]
> @iremyux can you please allow contributions from reviewers so that I can apply review suggestions, rebase and land the PR?

@malfet Done.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 8 above)
Comment (iremyux) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: zpotrs, cpotrs, dpotrs, spotrs, zunqmr, cunmqr, dormqr, and sormqr functions are not properly usable on Windows Arm64. Other platforms should be okay, hence the if guard for Windows Arm64.
We opened an issue on ARM Platform ( https://community.arm.com/support-forums/f/high-performance-computing-forum/56512/unable-to-use-lapack---potrs-functions) and mentioned it as a comment.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 9 above)
Comment (iremyux) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: We put a condition for Windows Arm64, to protect other platforms. So this endif is still necessary.

**Comment by ezyang** [comment]
@malfet you've got this one?

**Comment by malfet** [comment]
> you've got this one?

Yeah, I waited for some reference to upstream issue why this is needed

**Review by malfet — APPROVED** [review]
